doc/developer/design/20250321_password_authentication.md

```diff
@@ -0,0 +1,121 @@
+
+### 2. Password Management Syntax:
+Users should be able to manage their passwords and set passwords for roles they create. Superusers should be able to alter any non-internal user's password.
+- **Syntax Examples**:
+    - `CREATE ROLE hunter WITH PASSWORD 'hunter2';` 
+    - `ALTER ROLE hunter SET (PASSWORD 'hunter2');` 
+     - `ALTER USER user_name PASSWORD NULL`
+
+- **Password Policies**: We will not implement password policies but should enforce a configurable minimum password length. 
+
+- **Login Role Determination**: We should display whether a role has a password and can be used as a login role in `mz_roles`. See [pg_roles](https://www.postgresql.org/docs/9.1/view-pg-roles.html).
+
+### 3. Hashed Password Storage
+The hashed passwords must be stored in secure, read-only system tables in the catalog. This table will store the hashed password along with the associated user role. The passwords must not be stored in plain text. 
+
+- **Secure Hashing**: For future compatibility with SCRAM, passwords should be hashed using `scram-sha-256` encryption. The RFC for [SCRAM](https://datatracker.ietf.org/doc/html/rfc5802) dictates an iteration count of at least 4096. However, the RFC for SCRAM-SHA-256 dictates that "[the hash iteration-count should be such that a modern machine will take 0.1 seconds to perform the complete algorithm](https://www.rfc-editor.org/rfc/rfc7677)". This also dictates at least 4096 with a recommendation of 15000. We should aim for 15000 if it does not negatively impact connection times or load.
+
+- **Access Control**: Hashed passwords should not be extracted from the catalog and should not be accessible through queryable tables. Passwords must not be logged or output in stack traces.
+
+- **Password Dating**: The creation date of passwords will be recorded along with the hashed and salted password. When the password changes a creation date should be reset. The creation date of passwords should be queryable.
+
+- **OPTIONAL: Password Versioning**: Updates to the password hashing mechanisms may be required. As long as we are receiving passwords in plain text we should be able to take a validated password and replace the existing a new securely hashed value. This may require prefixing the password with data about the hash algorithm or parameters.
+
+### 4. Configurable System User Passwords:
+Passwords for `mz_system` and `mz_support` roles will be settable via environment variables `MZ_MZ_SYSTEM_ROLE_PASSWORD` and `MZ_MZ_SUPPORT_ROLE_PASSWORD`. Orchestratord should provide a set of parameters to set these variables via a kubernetes secret. Additionally, We should enable login of `mz_system` through the external ports when they have passwords set. This should require both an explicit feature flag, `enable_public_internal_user_login` to be set to true, and a password set in an environment variable.
```

Comment:
Setting these in an environment variable would work, but would require a restart to change the password. If we instead mount the secret as a file, we can use an inotify watch to trigger an immediate reload of the file.

I'm also not sure about the behavior of this feature flag. Wouldn't we want that to default to true? Do we even need a feature flag if we have a CLI argument for the file path of this system user secrets file?